### PR TITLE
Remove is_relationally_comparable for table device views

### LINKED ISSUE #10080: [BUG] cudaMalloc and cudaFree are being called during aggregations
**Describe the bug**
While examining a recent trace I noticed that within the libcudf `aggregate` range there are calls to `cudaMalloc` and `cudaFree`, the latter which causes a synchronization on the default stream.  I attached gdb and put a breakpoint on `cudaMalloc` and found it was being triggered by `cudf::detail::is_relationally_comparable<cudf::table_device_view>` because it calls `thrust::all_of` without passing an execution policy.  Without using the RMM policy, it will use the default CUDA allocator.  Ideally it should be using `rmm::exec_policy(stream)` but the stream is not available to this method and would need to be passed.

**Steps/Code to reproduce bug**
Attach a debugger to a query using the RMM arena allocator and executes an aggregation.  Place a breakpoint on `cudaMalloc` and execute the query and observe the breakpoint is hit in a callstack that derives from `cudf::detail::is_relationally_comparable`.

**Expected behavior**
libcudf should not trigger calls to `cudaMalloc` or `cudaFree`.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #10080

Removes `is_relationally_comparable` function for `table_device_view` and `mutable_table_device_view`. The existing `is_relationally_comparable` can still be used with non-device `table_view` and `mutable_table_view`. 

Callers are updated as appropriate. The incorrect kernel call has been removed.

### CODE REVIEW

File: cpp/src/table/table_device_view.cu [hunk 1]
```diff
@@ -86,13 +87,31 @@ extern template bool is_relationally_comparable<table_view>(table_view const& lh
...
+                                rmm::cuda_stream_view stream)
+{
+  return thrust::all_of(rmm::exec_policy(stream),
+                        thrust::counting_iterator<size_type>(0),
+                        thrust::counting_iterator<size_type>(lhs.num_columns()),
+                        [lhs, rhs] __device__(auto const i) {
+                          return lhs.column(i).type() == rhs.column(i).type() and
+                                 type_dispatcher(lhs.column(i).type(),
+                                                 is_relationally_comparable_functor{});
+                        });
```
Comment (jrhemstad) [inline cpp/src/table/table_device_view.cu:102]: Why does this need to run on device? All we're doing is looking at the types of the columns. There shouldn't need to be anything on the device here.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I am very confused about why this code needs a stream or run on the device at all.

File: cpp/src/table/table_device_view.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/table/table_device_view.cu:102]: The column array in a `table_device_view` is in device memory -- this call `lhs.column(i)` requires accessing device memory.

Comment (davidwendt) [reply cpp/src/table/table_device_view.cu:102]: https://github.com/rapidsai/cudf/blob/c1638869116aae2c6dde6024394279a2fb79e685/cpp/include/cudf/table/table_device_view.cuh#L57-L62

Comment (jrhemstad) [reply cpp/src/table/table_device_view.cu:102]: Oh, this is not good. The fact that it's a template hid that this is what's going on. 

This function should definitely not be launching a kernel.  Which means the function probably shouldn't exist for `table_device_view` and only `table_view`, which would require modifying the `row_comparator` here. 

I guess this fix is necessary and fine for now, but we should work to eliminate needing this ASAP.

Comment (davidwendt) [reply cpp/src/table/table_device_view.cu:102]: There seems to be no easy way for the `row_lexicographic_comparator` to check comparability here. If we pass non-device views on the [constructor](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/include/cudf/table/row_operators.cuh#L372-L376) instead then the check can be done efficiently in host code. But the comparator needs device-views for it's [`operator()`](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/include/cudf/table/row_operators.cuh#L397) method. If the `row_lexicographic_comparator` were to create the device-views in the constructor we would need to store unique-ptrs in the member data. This would delete the copy-constructor which is needed to pass the functor to [thrust calls for example](https://github.com/rapidsai/cudf/blob/87a2ea485ef9a91fb9df46eb6135f083671147cb/cpp/src/sort/is_sorted.cu#L50-L59). Storing the unique-ptrs as shared-ptrs in the member data would work but means that the comparator cannot be passed through to lower device functions since the copy-constructor would try to increment the shared-ptr reference counter which is stored in host memory.

I don't think we want to pass both the device and non-device table views to the constructor so my inclination here is to make the caller ensure comparability of the table columns before using the `row_lexicographic_comparator`. This would remove the need for a kernel launch (thrust device call) in the constructor.

@devavret Maybe this can be incorporated in your rework of the comparators?

**Comment by vyasr** [comment]
@davidwendt just FYI I have avoided reviewing this PR so far because I'm not clear on whether @jrhemstad's question is resolved or whether to expect a real resolution within the scope of this PR, and it seems like a decision there could significantly alter this changeset. Please let me know if you think I should just go ahead and review.

**Comment by davidwendt** [comment]
> @davidwendt just FYI I have avoided reviewing this PR so far because I'm not clear on whether @jrhemstad's question is resolved or whether to expect a real resolution within the scope of this PR, and it seems like a decision there could significantly alter this changeset. Please let me know if you think I should just go ahead and review.

Thanks @vyasr 
I think the right approach here is to remove the troublesome `is_relationally_comparable_functor` for table device-views. This will remove the incorrect kernel/thrust call and solve the original issue. I believe @devavret will be wrapping the `row_lexicographic_comparator` with an owning class in a separate PR (#10164) which could re-enable the comparable checking on non-device table views there instead.

**Comment by devavret** [comment]
> I believe @devavret will be wrapping the `row_lexicographic_comparator` with an owning class in a separate PR (#10164) which could re-enable the comparable checking on non-device table views there instead.

That is correct. The check will then be on `table_view`. But #10164 will have the new comparator in namespace experimental and won't immediately replace the existing comparator. So there may be benefit in this PR going through now.

File: cpp/include/cudf/table/row_operators.cuh [hunk 2]
```diff
@@ -355,12 +355,13 @@ class row_lexicographic_comparator {
    * @brief Construct a function object for performing a lexicographic
    * comparison between the rows of two tables.
    *
+   * The caller should ensure the corresponding column types are comparable.
```
Comment (vyasr) [inline cpp/include/cudf/table/row_operators.cuh:358]:
```suggestion
   * The caller should ensure the corresponding column types are comparable.
   * Behavior is undefined if called with incomparable column types.
```

Comment (vyasr) [reply cpp/include/cudf/table/row_operators.cuh:358]: I prefer to be explicit about UB.

File: cpp/src/table/table_device_view.cu [hunk 3]
```diff
@@ -20,8 +20,7 @@
 #include <cudf/utilities/error.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
-
-#include <thrust/logical.h>
+#include <rmm/exec_policy.hpp>
```
Comment (vyasr) [inline cpp/src/table/table_device_view.cu:23]: Why was this added? It doesn't look necessary in this file.

**Review by vyasr — APPROVED** [review]
LGTM.

**Review by jrhemstad — APPROVED** [review]

File: cpp/src/table/table_view.cpp [hunk 4]
```diff
@@ -96,4 +96,36 @@ table_view scatter_columns(table_view const& source,
...
+                                              is_relationally_comparable_functor{});
+                     });
+}
+
+// Explicit extern template instantiation for a table of immutable views
+extern template bool is_relationally_comparable<table_view>(table_view const& lhs,
+                                                            table_view const& rhs);
+
+// Explicit extern template instantiation for a table of mutable views
+extern template bool is_relationally_comparable<mutable_table_view>(mutable_table_view const& lhs,
```
Comment (ttnghia) [inline cpp/src/table/table_view.cpp:127]: `extern` should not be used here. As I found https://stackoverflow.com/questions/8130602/using-extern-template-c11):
```
You should only use extern template to force the compiler to not instantiate a template when you know that it will be instantiated somewhere else. It is used to reduce compile time and object file size.
```

**Review by ttnghia — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge
